### PR TITLE
fix(mayastor): nexus child devices must be registered before they can be closed

### PR DESCRIPTION
**Pull request by cjones1024**
Properly closing a nexus child requires an event listener to be registered
in order to handle the resulting device removal event.

There is a bug in Nexus::try_open_children(), in which a nexus child may be
opened and then closed again before any event listener has been registered,
causing the close() to hang.

The fix is to ensure that the event listener is registered
as soon as a child is successfully opened.

### CODE REVIEW

File: mayastor/src/bdev/nexus/nexus_bdev.rs [hunk 1]
```diff
@@ -883,11 +886,17 @@ pub async fn nexus_create(
 }
 
 /// Destroy list of child bdevs
-async fn destroy_child_bdevs(name: &str, list: &[String]) {
-    let futures = list.iter().map(String::as_str).map(bdev_destroy);
+async fn destroy_child_bdevs(nexus: &str, list: &[String]) {
+    let futures = list
+        .iter()
+        .map(String::as_str)
+        .filter(|name| Bdev::lookup_by_name(name).is_some())
```
Comment (mtzaurus) [inline mayastor/src/bdev/nexus/nexus_bdev.rs:893]: Does this properly find our non-bdev NVMx devices ? IIRC, Bdev::lookup_by_name() ends up in 'spdk_bdev_get_by_name()', which can't see our NVMe controllers.

Comment (cjones1024) [reply mayastor/src/bdev/nexus/nexus_bdev.rs]: Have replaced with device_lookup(), and also used device_destroy().
There no longer seems to be any valid use for bdev_destroy(), and that should probably be removed.

**Review by gila — CHANGES_REQUESTED** [review]
I'd like to see a test case for this, it should have been there already ofcourse, but A test case where the nexus is destroyed properly when a child fails to connect would help.

**Comment by cjones1024** [comment]
> 
> 
> I'd like to see a test case for this, it should have been there already ofcourse, but A test case where the nexus is destroyed properly when a child fails to connect would help.

This scenario is most easily triggered if a child fails to open by virtue of being too small for the specified nexus size. Have now added a new cargo test for this.

**Pull request closed without merging**